### PR TITLE
Kalyn - L115.nh3_an_R_S_T_Y

### PR DESCRIPTION
**Pull request by kdorheim**
dsr for the module_emissions_L115.nh3_an_R_S_T_Y  chunk. This chunk
computes annual animal NH3 emissions by GCAM region, sector and
technology. It passes the test package but produces a weird document
warning with check package.

#355

### CODE REVIEW

**Comment by bpbond** [comment]
Hi @kdorheim - thanks! Let's sit down tomorrow and resolve these conflicts.

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 1]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+
+    # Match emissions factors by year and compute unscaled emissions
+    L115.nh3_tg_R_an_C_Sys_Fd_Yh %>%
+      left_join(L105.nh3_tgmt_USA_an_Yh, by = "year") %>%
+      rename(emfact = value) %>%
+      mutate(epa_emissions = production * emfact) %>%
+      na.omit() ->
+      L115.nh3_tg_R_an_C_Sys_Fd_Yh.mlt
+
+    # Match by sector, technology and fuel
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: A little more detailed comment would be useful here

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 2]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      summarize(EPA_emissions = sum(epa_emissions)) %>%
+      arrange(year) ->
+      L115.nh3_tg_R_an_C_Yh.mlt
+
+    # Compute EDGAR emissions by region and sector
+    L115.EDGAR <- EDGAR_NH3
+
+    # Renaming L115.EDGAR variable names - Note: may remove if names for L115.EDGAR change
+    names(L115.EDGAR)[1] <- "IPCC_Annex"
+    names(L115.EDGAR)[2] <- "World_Region"
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: Why not use `rename` for these lines, for consistency?

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 3]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      mutate(Non.CO2 = "NH3_AGR") %>%
+      left_join(EDGAR_sector, by = "IPCC") %>%
+      rename(EDGAR_agg_sector = agg_sector) ->
+      L115.EDGAR
+
+    # Convert from EDGAR iso to GCAM_region_ID
+    L115.EDGAR %>%
+      mutate( iso = tolower( ISO_A3 ), ISO_A3 = NULL ) %>%
+      change_iso_code('rou', 'rom') %>%
+      left_join(iso_GCAM_regID, by = "iso") ->
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: Should be `left_join_error_no_match`? In general (e.g. lines 56, 80, 106, 114) if you can use a more restrictive join, please do so.

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 4]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      L115.EDGAR
+
+    # Note : the old data system drops values from 1970, it is unclear if that is intentional or an accident.
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: Did you check with Page about this?

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 5]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      tidyr::separate(Region_GHG_Sector_Yr, c("GCAM_region_ID", "Non.CO2", "EDGAR_agg_sector", "year"), sep = "~") %>%
+      mutate( ., scaler = EDGAR_emissions / EPA_emissions / 1000.0 ) ->
+      L115.emiss_scaler
+
+    # Now, match by GCAM Region ID, gas, EDGAR_agg_sector and year, then scale EPA emission by tech to match EDGAR tools
+    tidyr::unite(L115.nh3_tg_R_an_C_Sys_Fd_Yh.mlt, col = "Region_GHG_Sector_Yr", c(GCAM_region_ID, Non.CO2, EDGAR_agg_sector, year), sep = "~") %>%
+      left_join(tidyr::unite(L115.emiss_scaler, col = "Region_GHG_Sector_Yr", c(GCAM_region_ID, Non.CO2, EDGAR_agg_sector, year), sep = "~"), by = "Region_GHG_Sector_Yr") %>%
+      tidyr::separate(Region_GHG_Sector_Yr, c("GCAM_region_ID", "Non.CO2", "EDGAR_agg_sector", "year"), sep = "~") %>%
+      mutate(emissions = epa_emissions * scaler) %>%
+      mutate_all(funs(replace(., is.na(.), 0))) ->
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: ?   is the data frame wide at this point? What about using `replace_na`?

File: R/zchunk_L115.nh3_an_R_S_T_Y.R [hunk 6]
```diff
@@ -31,47 +29,112 @@ module_emissions_L115.nh3_an_R_S_T_Y_DISABLED <- function(command, ...) {
...
+      L115.nh3_tg_R_an_C_Sys_Fd_Yh.mlt
+
+    # Rename columns and reshape data
+    L115.nh3_tg_R_an_C_Sys_Fd_Yh.mlt %>%
+      rename(supplysector = GCAM_commodity, subsector = system, stub.technology = feed, value = emissions) %>%
```
Comment (bpbond) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: `select` can also be used to rename, FYI; could combine lines 122 and 123 into a single `select` I think.

**Comment by bpbond** [comment]
Not sure why this closed. Weird. Re-opening.

**Comment by kdorheim** [comment]
Hi @bpbond - sounds good, I am around all day until 4:30.

File: R/zchunk_L115.nh3_an_R_S_T_Y.R (hunk 5 above)
Comment (kdorheim) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: `replace_na(list(emissions = 0))`

File: R/zchunk_L115.nh3_an_R_S_T_Y.R (hunk 4 above)
Comment (kdorheim) [inline R/zchunk_L115.nh3_an_R_S_T_Y.R]: Yes sorry I thought I had removed this.

**Comment by bpbond** [comment]
@ssmithClimate This is also in your review queue if possible. Thanks!

**Comment by bpbond** [comment]
@kdorheim make sure the package builds before you push changes please! I've resolved conflicts.

**Comment by bpbond** [comment]
@kdorheim This now has problems, throwing a `left_join_error_no_match` error.

**Comment by kdorheim** [comment]
@bpbond one of your initial comments with the chunk was that I did not used `left_join_error_no_match` but when I do the driver fails. In that case it is okay to use `left_join`?

**Comment by bpbond** [comment]
Yes! The whole point of left `left_join_error_no_match` is that you're asserting "I'm doing a join here for which there should be a 1-to-1 lookup, i.e. no NAs or multiple matches". If that's not true, don't use it (but explain why in a comment). Make sense?

**Comment by bpbond** [comment]
Assuming this builds w/o error on Travis, looks good. Waiting on @ssmithClimate approval to merge.

**Comment by bpbond** [comment]
OK to merge @ssmithClimate ?

**Pull request merged**